hipcc: Tell hipcc where to find the various tools it depends on
FindHIP: Improved search logic for HIP_ROOT_DIR
diff --git a/bin/hipcc.pl b/bin/hipcc.pl
index 645ae62d..f26f1262 100755
--- a/bin/hipcc.pl
+++ b/bin/hipcc.pl
@@ -99,12 +99,15 @@ $HIP_RUNTIME    =   $hipvars::HIP_RUNTIME;
 $HIP_PLATFORM   =   $hipvars::HIP_PLATFORM;
 $HIP_COMPILER   =   $hipvars::HIP_COMPILER;
 $HIP_CLANG_PATH =   $hipvars::HIP_CLANG_PATH;
+$HIP_CLANG_RUNTIME = $hipvars::HIP_CLANG_RUNTIME;
 $CUDA_PATH      =   $hipvars::CUDA_PATH;
 $HIP_PATH       =   $hipvars::HIP_PATH;
 $ROCM_PATH      =   $hipvars::ROCM_PATH;
+$ROCMINFO_PATH  =   $hipvars::ROCMINFO_PATH;
 $HIP_VERSION    =   $hipvars::HIP_VERSION;
-$HSA_PATH       =   $hipvars::HSA_PATH;
 $HIP_ROCCLR_HOME =   $hipvars::HIP_ROCCLR_HOME;
+$DEVICE_LIB_PATH =   $hipvars::DEVICE_LIB_PATH;
+$HSA_PATH       =   "$ROCM_PATH/hsa";
 
 if ($HIP_PLATFORM eq "amd") {
   # If using ROCclr runtime, need to find HIP_ROCCLR_HOME
@@ -204,7 +207,6 @@ if ($HIP_PLATFORM eq "amd") {
     }
 
     if (not $isWindows) {
-        $HSA_PATH=$ENV{'HSA_PATH'} // "$ROCM_PATH/hsa";
         $HIPCXXFLAGS .= " -isystem $HSA_PATH/include";
         $HIPCFLAGS .= " -isystem $HSA_PATH/include";
     }
@@ -612,7 +614,7 @@ if($HIP_PLATFORM eq "amd"){
             $targetsStr = $ENV{HCC_AMDGPU_TARGET};
         } elsif (not $isWindows) {
             # Else try using rocm_agent_enumerator
-            $ROCM_AGENT_ENUM = "${ROCM_PATH}/bin/rocm_agent_enumerator";
+            $ROCM_AGENT_ENUM = "${ROCMINFO_PATH}/bin/rocm_agent_enumerator";
             $targetsStr = `${ROCM_AGENT_ENUM} -t GPU`;
             $targetsStr =~ s/\n/,/g;
         }
@@ -716,11 +718,11 @@ if ($HIP_PLATFORM eq "amd") {
         $toolArgs = ${toolArgs} . " -Wl,--enable-new-dtags -Wl,-rpath=$HIP_LIB_PATH:$ROCM_PATH/lib -lamdhip64 ";
       }
       # To support __fp16 and _Float16, explicitly link with compiler-rt
-      $HIP_CLANG_BUILTIN_LIB="$HIP_CLANG_PATH/../lib/clang/$HIP_CLANG_VERSION/lib/$HIP_CLANG_TARGET/libclang_rt.builtins.a";
+      $HIP_CLANG_BUILTIN_LIB="$HIP_CLANG_RUNTIME/lib/linux/libclang_rt.builtins.a";
       if (-e $HIP_CLANG_BUILTIN_LIB) {
-        $toolArgs .= " -L$HIP_CLANG_PATH/../lib/clang/$HIP_CLANG_VERSION/lib/$HIP_CLANG_TARGET -lclang_rt.builtins "
+        $toolArgs .= " -L$HIP_CLANG_RUNTIME/lib/linux -lclang_rt.builtins "
       } else {
-        $toolArgs .= " -L$HIP_CLANG_PATH/../lib/clang/$HIP_CLANG_VERSION/lib/linux -lclang_rt.builtins-x86_64 "
+        $toolArgs .= " -L$HIP_CLANG_RUNTIME/lib/linux -lclang_rt.builtins-x86_64 "
       }
     }
 }
@@ -749,6 +751,8 @@ if ($needLDFLAGS and not $compileOnly) {
 }
 $CMD .= " $toolArgs";
 
+$CMD .= " --hip-path=${HIP_PATH}";
+
 if ($verbose & 0x1) {
     print "hipcc-cmd: ", $CMD, "\n";
 }
diff --git a/cmake/FindHIP.cmake b/cmake/FindHIP.cmake
index 8b196dea..ac0baad5 100644
--- a/cmake/FindHIP.cmake
+++ b/cmake/FindHIP.cmake
@@ -66,6 +66,9 @@ if(NOT APPLE)
             DOC "HIP installed location"
             NO_DEFAULT_PATH
             )
+        if(NOT HIP_ROOT_DIR)
+            find_path(HIP_ROOT_DIR NAMES bin/hipconfig)
+        endif()
         if(NOT EXISTS ${HIP_ROOT_DIR})
             if(HIP_FIND_REQUIRED)
                 message(FATAL_ERROR "Specify HIP_ROOT_DIR")
